restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -242,7 +242,7 @@ public static RestException buildRestException(RestLiResponseException restLiRes
         .setStatus(restLiResponse.getStatus() == null ? HttpStatus.S_500_INTERNAL_SERVER_ERROR.getCode()
             : restLiResponse.getStatus().getCode());
 
-    if (restLiResponse.hasData())
+    if (restLiResponse.hasData() && restLiResponse.getStatus() != HttpStatus.S_204_NO_CONTENT)
```

Comment:
This place meant for the major point of this fix

Reply:
Thanks for the quick fix. One more thing we need to do is to log error if an exception is thrown with success status code. 

I thought doing this in error response builder would help with the logging:

```
diff --git a/restli-server/src/main/java/com/linkedin/restli/internal/server/response/ErrorResponseBuilder.java b/restli-server/src/main/java/com/linkedin/restli/internal/server/response/ErrorResponseBuilder.java
index 71700b6c3..1268439d3 100644
--- a/restli-server/src/main/java/com/linkedin/restli/internal/server/response/ErrorResponseBuilder.java
+++ b/restli-server/src/main/java/com/linkedin/restli/internal/server/response/ErrorResponseBuilder.java
@@ -16,10 +16,9 @@
 
 package com.linkedin.restli.internal.server.response;
 
-import com.linkedin.data.schema.RecordDataSchema;
-import com.linkedin.data.template.RecordTemplate;
 import com.linkedin.restli.common.ErrorDetails;
 import com.linkedin.restli.common.ErrorResponse;
+import com.linkedin.restli.common.HttpStatus;
 import com.linkedin.restli.common.ProtocolVersion;
 import com.linkedin.restli.common.ResourceMethod;
 import com.linkedin.restli.common.RestConstants;
@@ -29,12 +28,13 @@ import com.linkedin.restli.internal.server.RoutingResult;
 import com.linkedin.restli.server.ErrorResponseFormat;
 import com.linkedin.restli.server.RestLiResponseData;
 import com.linkedin.restli.server.RestLiServiceException;
-
 import java.io.PrintWriter;
 import java.io.StringWriter;
 import java.net.HttpCookie;
 import java.util.List;
 import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
@@ -43,6 +43,7 @@ import java.util.Map;
  */
 public final class ErrorResponseBuilder
 {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ErrorResponseBuilder.class.getName());
   public static final String DEFAULT_INTERNAL_ERROR_MESSAGE = "INTERNAL SERVER ERROR";
   private final ErrorResponseFormat _errorResponseFormat;
 
@@ -72,6 +73,18 @@ public final class ErrorResponseBuilder
 
   public ErrorResponse buildErrorResponse(RestLiServiceException result)
   {
+    if (result.getStatus() != null && result.getStatus().getCode() < 400)
+    {
+      // Invalid to send an error response with success status codes. This should be converted to 500 errors.
+      // Logging an error message now to detect and fix current use cases before we start converting to 500.
+      LOGGER.error("Incorrect use of success status code with error response", result);
+    }
+
+    if (result.getStatus() == HttpStatus.S_204_NO_CONTENT)
+    {
+      // HTTP Spec requires the response body to be empty for HTTP status 204.
+      return null;
+    }
     return buildErrorResponse(result, result.hasOverridingErrorResponseFormat() ? result.getOverridingFormat() : _errorResponseFormat);
   }
```

Reply:
We can still keep your change, it can guard against success responses with 204 status code.

Reply:
I merged yours within my PR